common/kokkos-sampler/kp_sampler.cpp

```diff
@@ -0,0 +1,192 @@
+	const uint64_t interfaceVer,
+	const uint32_t devInfoCount,
+	void* deviceInfo) {
+
+	const char* tool_verbose_str = getenv("KOKKOS_SAMPLER_VERBOSE");
+	if( NULL != tool_verbose_str ) {
+		tool_verbosity = atoi( tool_verbose_str );
+	} else {
+		tool_verbosity = 0;
+	}
+
+	char* profileLibrary = getenv("KOKKOS_PROFILE_LIBRARY");
+	char* envBuffer = (char*) malloc( sizeof(char) * (strlen(profileLibrary) + 1 ));
+	sprintf(envBuffer, "%s", profileLibrary);
+
+	char* nextLibrary = strtok(envBuffer, ";");
+
+	for(int i = 0; i < loadSeq; i++) {
+		nextLibrary = strtok(NULL, ";");
+	}
+
+	nextLibrary = strtok(NULL, ";");
+
+	if(NULL == nextLibrary) {
+		printf("KokkosP: No child library to call in %s\n", profileLibrary);
+		exit(-1);
+	} else {
+		if( tool_verbosity > 0 ) {
+			printf("KokkosP: Next library to call: %s\n", nextLibrary);
+			printf("KokkosP: Loading child library ..\n");
+		}
+
+		void* childLibrary = dlopen(nextLibrary, RTLD_NOW | RTLD_GLOBAL);
+
+		if(NULL == childLibrary) {
+			fprintf(stderr, "KokkosP: Error: Unable to load: %s (Error=%s)\n",
+				nextLibrary, dlerror());
+			exit(-1);
+		} else {
+			beginForCallee = (beginFunction) dlsym(childLibrary, "kokkosp_begin_parallel_for");
+	               	beginScanCallee = (beginFunction) dlsym(childLibrary, "kokkosp_begin_parallel_scan");
+	               	beginReduceCallee = (beginFunction) dlsym(childLibrary, "kokkosp_begin_parallel_reduce");
+
+        	        endScanCallee = (endFunction) dlsym(childLibrary, "kokkosp_end_parallel_scan");
+       	        	endForCallee = (endFunction) dlsym(childLibrary, "kokkosp_end_parallel_for");
+    	           	endReduceCallee = (endFunction) dlsym(childLibrary, "kokkosp_end_parallel_reduce");
+
+   	            	initProfileLibrary = (initFunction) dlsym(childLibrary, "kokkosp_init_library");
+	               	finalizeProfileLibrary = (finalizeFunction) dlsym(childLibrary, "kokkosp_finalize_library");
+
+			if(NULL != initProfileLibrary) {
+				(*initProfileLibrary)(loadSeq + 1, interfaceVer, devInfoCount, deviceInfo);
+			}
+
+			if( tool_verbosity > 0 ) {
+				printf("KokkosP: Function Status:\n");
+				printf("KokkosP: begin-parallel-for:      %s\n", (beginForCallee == NULL) ? "no" : "yes");
+				printf("KokkosP: begin-parallel-scan:     %s\n", (beginScanCallee == NULL) ? "no" : "yes");
+				printf("KokkosP: begin-parallel-reduce:   %s\n", (beginReduceCallee == NULL) ? "no" : "yes");
+				printf("KokkosP: end-parallel-for:        %s\n", (endForCallee == NULL) ? "no" : "yes");
+				printf("KokkosP: end-parallel-scan:       %s\n", (endScanCallee == NULL) ? "no" : "yes");
+				printf("KokkosP: end-parallel-reduce:     %s\n", (endReduceCallee == NULL) ? "no" : "yes");
+			}
+		}
+	}
+
+	free(envBuffer);
+
+	kernelCounter = 0;
+	uniqID = 1;
+
+	const char* tool_sample_rate = getenv("KOKKOS_SAMPLER_RATE");
+	if( NULL != tool_sample_rate ) {
+		kernelSampleRate = atoi( tool_sample_rate );
+	} else {
+		kernelSampleRate = 101;
+	}
```

Comment:
Actually now that I built the tool and gave it a spin, I find this counterintuitive. If I don't look at the implementation, when I select a "sampling rate" 2x bigger, I would expect to record twice more data but this does the opposite. 
Currently, `KOKKOS_SAMPLER_RATE=1` means record everything, `KOKKOS_SAMPLER_RATE=10` record every 10 kernels.

Reply:
Yeah its not a rate I guess.

Reply:
maybe rename to KOKKOS_SAMPLER_EVERY=10? As in "every 10th"? Or you could read in a fraction.